### PR TITLE
URL prefetching, Ads preconnect and prefetching and Safari preconnect

### PR DESCRIPTION
**Pull request by cramforce**
- Implements URL prefetching and uses it in a few places where it makes sense.
- Adds the capability for ad networks to preconnect and prefetch resources
before being actually loaded.
- Adds a preconnect polyfill, mostly for Safari, that should speed up
3p embeds on Safari.
- Optimizes prefetch for 3p iframe to correctly prefetch custom bootstrap
iframes	and to also prefetch the 3p frame base script.

### CODE REVIEW

File: ads/_prefetch.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/**
```
Comment (dvoytenko) [inline ads/_prefetch.js]: Why underscore in file?

File: builtins/amp-ad.js [hunk 2]
```diff
@@ -142,6 +142,37 @@ export function installAd(win) {
...
+      }
+      if (typeof preconnect == 'string') {
+        this.preconnect.url(preconnect);
+      } else if (preconnect) {
+        preconnect.forEach(p => {
+          this.preconnect.url(p);
+        });
+      }
+      // If fully qualified src for ad script is specified we prefetch that.
+      var src = this.element.getAttribute('src');
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: let

File: builtins/amp-ad.js [hunk 3]
```diff
@@ -142,6 +142,37 @@ export function installAd(win) {
     }
 
     /**
+     * Prefetches and preconnects URLs related to the ad.
+     * @private
+     */
+    prefetchAd_() {
+      // We always need the bootstrap.
+      prefetchBootstrap(this.getWin());
+      var type = this.element.getAttribute('type');
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: let here and elsewhere

File: ads/_prefetch.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * URLs to prefetch for a given ad type.
+ *
+ * This MUST be kept in sync with actual implementation.
+ *
+ * @const {string, (string|!Array<string>)}
```
Comment (dvoytenko) [inline ads/_prefetch.js]: !Object?

**Comment by dvoytenko** [comment]
LGTM with a couple of comments.

File: ads/_prefetch.js (hunk 1 above)
Comment (cramforce) [inline ads/_prefetch.js]: Maybe a bad idea. I kinda wanted this to stand out since it needs to be updated a lot.

File: builtins/amp-ad.js (hunk 2 above)
Comment (cramforce) [inline builtins/amp-ad.js]: Done.

File: builtins/amp-ad.js (hunk 3 above)
Comment (cramforce) [inline builtins/amp-ad.js]: Done

File: ads/_prefetch.js (hunk 4 above)
Comment (cramforce) [inline ads/_prefetch.js]: Indeed

**Pull request merged**